### PR TITLE
Amazon Offering Parsing

### PR DESCRIPTION
**Pull request by beylmk**
Amazon only has a single productId (the `termSku`) returned in offerings, while google has both the subId and the basePlanID. So now the way we parse offerings must split between the stores.

This PR introduces an `OfferingParser` and `OfferingParserFactory`, instantiating the correct parser from the factory upon initialization. I'd love to get some feedback on this approach before writing tests, as I think those will be a lot of work...

My thoughts on other potential implementations and why i chose this one:

* simply pass the store as a parameter into the OfferingFactories methods, and if/else split how we do parsing on that 

this is the easiest, but it breaks our general "don't have store-specific stuff in purchases" rule. see [this commit](00e22606b5938daf8e39646868c59e31a08de2ca) as an example

* make OfferingFactory an abstract class and instantiate a store-specific implementation upon purchases initialization, somewhat like BillingFactory works. see [this commit](ce110fdaeecdb2da7ff711501a7ad1f6f0528658) for example (it's not complete/perfect, i just committed for easy reviewing of the overall architecture idea)

this feels a bit like over-architecting to me... but i could see more complex offering parsing in the future -- for instance, if we implement sorted `SubscriptionOptions`. This is the route I took in this PR.

* pull the offering parsing into Billing implementation. There is really only one line of code that differs, so we could only pull line out into a function in the Billing classes for now (and keep the rest in the base billing class). 

I went this route [here](https://github.com/RevenueCat/purchases-android/pull/802), but it doesn't feel totally right since offering parsing doesn't communicate with the billing libraries and isn't really a "billing" thing.

### CODE REVIEW

File: common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt [hunk 1]
```diff
@@ -39,7 +40,8 @@ class OfferingsTest {
 
         val packageToTest = getPackageJSON().createPackage(
             products,
-            "offering"
+            "offering",
+            Store.PLAY_STORE
```
Comment (swehner) [inline common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt:44]: All these changes are completely unrelated, right?

File: common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    private fun createPackage(
+        packageJson: JSONObject,
+        productsById: Map<String, List<StoreProduct>>,
+        offeringIdentifier: String
+    ): Package? {
+        val packageIdentifier = packageJson.getString("identifier")
+        val productIdentifier = packageJson.getString("platform_product_identifier")
+
+        val planIdentifier = packageJson.optString("platform_product_plan_identifier").takeIf { it.isNotEmpty() }
+        val product = findMatchingProduct(productsById, productIdentifier, planIdentifier)
```
Comment (swehner) [inline common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt:69]: I think this interface is a bit strange - one difference between Amazon and Google is that Amazon just needs `platform_product_identifier` and that Google also needs the `platform_product_plan_identifier`. I would move this difference into the `findMatchingProduct` method.
Otherwise for any new piece of information we'll have to change the API again.

Maybe we can do `findMatchingProduct(productsById, packageJson)`?

**Review by swehner — COMMENTED** [review]
Looks good - I like the approach. Small comment re: the interface

File: common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt (hunk 1 above)
Comment (beylmk) [reply common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt:44]: oh yeah, the test changes are left from a different approach. was waiting to update them to this one until i get sign off

File: common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt (hunk 2 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt:69]: fair point... i was just trying to reuse as much code as possible but it's less future-proof.

File: common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import com.revenuecat.purchases.Offerings
+import com.revenuecat.purchases.Package
+import com.revenuecat.purchases.PackageType
+import com.revenuecat.purchases.models.StoreProduct
+import com.revenuecat.purchases.strings.OfferingStrings
+import org.json.JSONObject
+
+abstract class OfferingParser {
+
+    abstract fun findMatchingProduct(
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt:13]: This could be `protected` I think?

File: common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        val planIdentifier = packageJson.optString("platform_product_plan_identifier").takeIf { it.isNotEmpty() }
+        val product = findMatchingProduct(productsById, productIdentifier, planIdentifier)
+
+        val packageType = packageIdentifier.toPackageType()
+        return product?.let { Package(packageIdentifier, packageType, it, offeringIdentifier) }
+    }
+}
+
+fun String.toPackageType(): PackageType =
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt:76]: And this could be private I think?

File: common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt (hunk 2 above)
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/OfferingParser.kt:69]: Yeah, since we are trying to avoid having store-specific logic here, that makes sense.

**Review by tonidero — COMMENTED** [review]
Sorry for the delay! The approach makes sense to me 👍

**Comment by beylmk** [comment]
@tonidero thanks for the review :)

**Review by vegaro — COMMENTED** [review]
Approach also makes sense to me. Sorry for the delay

File: feature/google/src/main/java/com/revenuecat/purchases/google/GoogleOfferingParser.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        if (planIdentifier == null) {
+            // It could be an INAPP or a mis-configured subscription
+            // Try to find INAPP, otherwise null
+            return storeProducts
+                .takeIf { it?.size == 1 }
+                ?.takeIf { it[0].type == ProductType.INAPP }
+                ?.firstOrNull()
+        }
+        return storeProducts?.firstOrNull { storeProduct ->
+            storeProduct.subscriptionOptions?.firstOrNull { it.isBasePlan }?.id == planIdentifier
```
[the remark this answers is not in the document]
Comment (joshdholtz) [reply feature/google/src/main/java/com/revenuecat/purchases/google/GoogleOfferingParser.kt:26]: Yup yup, I think this is the only part of the improvements that can be used here 😇 Pretty small improvement but 🤷‍♂️

File: feature/amazon/src/test/java/com/revenuecat/purchases/amazon/AmazonOfferingsTest.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+//  Copyright © 2019 RevenueCat, Inc. All rights reserved.
+//
+
+package com.revenuecat.purchases.amazon
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.junit.runner.RunWith
+import org.robolectric.annotation.Config
+
+//@RunWith(AndroidJUnit4::class)
```
Comment (beylmk) [inline feature/amazon/src/test/java/com/revenuecat/purchases/amazon/AmazonOfferingsTest.kt:12]: i started writing these but felt like it would kinda be wasted time without AmazonStoreProduct work done..

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt [hunk 7]
```diff
@@ -4447,6 +4449,33 @@ class PurchasesTest {
...
+//            } just runs
+//            every {
+//                updatedCustomerInfoListener = any()
+//            } just runs
+//            every {
+//                updatedCustomerInfoListener
+//            } returns null
+//        }
+//    }
+
```
Comment (joshdholtz) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:4478]: Do these need to be uncommented or rewritten yet?

Comment (joshdholtz) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:4478]: Do these need to be uncommented or rewritten yet?

**Review by joshdholtz — APPROVED** [review]
This looks good to me! Just one question about some commented out tests and I think you have a commented about writing more tests but... LGTM 💪

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt (hunk 7 above)
Comment (beylmk) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:4478]: was gonna wait until amazonstoreproduct is done!
